{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/biodata/miniconda3/envs/cplex_aman/lib/python3.10/site-packages/cobamp/wrappers/external_wrappers.py:9: UserWarning: \n",
      "The wrappers.external_wrappers module will be deprecated in a future release in favour of the wrappers module. \n",
      "    Available ModelObjectReader classes can still be loaded using cobamp.wrappers.<class>. An appropriate model \n",
      "    reader can also be created using the get_model_reader function on cobamp.wrappers\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import cobra\n",
    "import re\n",
    "\n",
    "from troppo.omics.readers.generic import TabularReader\n",
    "from troppo.methods_wrappers import ModelBasedWrapper, ReconstructionWrapper\n",
    "from troppo.omics.integration import ContinuousScoreIntegrationStrategy\n",
    "from troppo.methods.reconstruction.gimme import GIMME, GIMMEProperties\n",
    "patt = re.compile('__COBAMPGPRDOT__[0-9]{1}')\n",
    "replace_alt_transcripts = lambda x: patt.sub('', x)\n",
    "# def replace_alt_transcripts(x):\n",
    "#     # remove COBAMP transcript suffixes like __COBAMPGPRDOT__12\n",
    "#     x = re.sub(r'__COBAMPGPRDOT__\\d+', '', x)\n",
    "#     # remove leading underscores before ENSG IDs\n",
    "#     x = re.sub(r'_(?=ENSG)', '', x)\n",
    "#     # insert missing spaces around 'and'/'or' if theyâ€™re glued to gene IDs\n",
    "#     x = re.sub(r'(?i)([A-Z0-9])(?=and)', r'\\1 ', x)\n",
    "#     x = re.sub(r'(?i)(and)(?=[A-Z0-9])', r'\\1 ', x)\n",
    "#     x = re.sub(r'(?i)([A-Z0-9])(?=or)', r'\\1 ', x)\n",
    "#     x = re.sub(r'(?i)(or)(?=[A-Z0-9])', r'\\1 ', x)\n",
    "#     # ensure lowercase boolean operators and single spaces\n",
    "#     x = re.sub(r'\\bAND\\b', 'and', x)\n",
    "#     x = re.sub(r'\\bOR\\b', 'or', x)\n",
    "#     x = re.sub(r'\\s+', ' ', x).strip()\n",
    "#     return x\n",
    "# load model and expression data\n",
    "model = cobra.io.read_sbml_model('/home/biodata/aman/Human-GEM/model/Human-GEM.xml')\n",
    "expression_data = pd.read_csv('/home/biodata/aman/expression_data_ensembl.csv', index_col=0)\n",
    "print(expression_data.head())\n",
    "\n",
    "expression_data_transposed = expression_data.T\n",
    "\n",
    "print(\"\\nTransposed shape:\", expression_data_transposed.shape)\n",
    "print(expression_data_transposed.head())\n",
    "\n",
    "omics_container = TabularReader(path_or_df=expression_data_transposed, \n",
    "                                nomenclature='gene',\n",
    "                                omics_type='transcriptomics').to_containers()\n",
    "\n",
    "single_sample = omics_container[0]\n",
    "model_wrapper = ReconstructionWrapper(model=model, ttg_ratio=9999,\n",
    "                                      gpr_gene_parse_function=replace_alt_transcripts)\n",
    "\n",
    "data_map = single_sample.get_integrated_data_map(model_reader=model_wrapper.model_reader,\n",
    "                                                 and_func=min, or_func=sum)\n",
    "print(list(data_map.get_scores().items())[:10])\n",
    "def score_apply(reaction_map_scores):\n",
    "    return {k:0  if v is None else v for k, v in reaction_map_scores.items()}\n",
    "\n",
    "continuous_integration = ContinuousScoreIntegrationStrategy(score_apply=score_apply)\n",
    "scores = continuous_integration.integrate(data_map=data_map)\n",
    "model = model_wrapper.model_reader.model\n",
    "\n",
    "# search for biomass in reaction names\n",
    "for r in model.reactions:\n",
    "    if 'med' in r.name.lower() or 'growth' in r.name.lower():\n",
    "        print(r.id, r.name)\n",
    "# print(model_wrapper.model_reader.r_ids)\n",
    "# Get the index of the biomass reaction in the model. This will be used as objective for the GIMME algorithm.\n",
    "idx_objective = model_wrapper.model_reader.r_ids.index('MAR13082')\n",
    "# Create the properties for the GIMME algorithm.\n",
    "properties = GIMMEProperties(exp_vector=[v for k, v in scores.items()], obj_frac=0.1, objectives=[{idx_objective: 1}],\n",
    "                             preprocess=True, flux_threshold=0.25, solver='CPLEX',\n",
    "                             reaction_ids= model_wrapper.model_reader.r_ids, metabolite_ids=model_wrapper.model_reader.m_ids)\n",
    "\n",
    "# Run the GIMME algorithm.\n",
    "gimme = GIMME(S=model_wrapper.S, lb=model_wrapper.lb, ub=model_wrapper.ub, properties=properties)\n",
    "\n",
    "gimme_run = gimme.run()\n",
    "print(type(gimme_run))\n",
    "print(gimme_run)\n",
    "pd.DataFrame(gimme_run).to_csv('gimme_output_tinit.tsv', sep='\\t')\n",
    "#---\n",
    "# model.reactions\n",
    "# len(model.reactions)\n",
    "selected_reactions = [model.reactions[i] for i in gimme_run]\n",
    "print(selected_reactions)\n",
    "#---\n",
    "ctx_model = model.copy()\n",
    "\n",
    "selected_ids = [r.id for r in selected_reactions]\n",
    "to_remove = [r for r in ctx_model.reactions if r.id not in selected_ids]\n",
    "\n",
    "ctx_model.remove_reactions(to_remove, remove_orphans=True)\n",
    "\n",
    "print(ctx_model)\n",
    "print(len(model.reactions), len(ctx_model.reactions))\n",
    "# export\n",
    "cobra.io.write_sbml_model(ctx_model, \"gimme_context_specific_model.xml\")\n",
    "solution = ctx_model.optimize()\n",
    "print(solution.objective_value)\n",
    "print(ctx_model.summary())\n",
    "fluxes = gimme.sol.__dict__['_Solution__value_map']\n",
    "print(fluxes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cplex_aman",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
